float32[4] position 